src/server/frontend_wayland/wlr_screencopy_v1.cpp

```diff
@@ -0,0 +1,252 @@
+        {
+            extents = config.extents();
+        }))
+    {
+        BOOST_THROW_EXCEPTION(std::runtime_error(
+            "No output for wl_output@" + std::to_string(wl_resource_get_id(output))));
+    }
+    return extents;
+}
+
+mf::WlrScreencopyManagerV1Global::WlrScreencopyManagerV1Global(
+    wl_display* display,
+    std::shared_ptr<WlrScreencopyV1Ctx> const& ctx)
+    : Global{display, Version<3>()},
+      ctx{ctx}
+{
+}
+
+void mf::WlrScreencopyManagerV1Global::bind(wl_resource* new_resource)
+{
+    new WlrScreencopyManagerV1{new_resource, ctx};
+}
+
+mf::WlrScreencopyManagerV1::WlrScreencopyManagerV1(
+    wl_resource* resource,
+    std::shared_ptr<WlrScreencopyV1Ctx> const& ctx)
+    : wayland::WlrScreencopyManagerV1{resource, Version<3>()},
+      ctx{ctx}
+{
+}
+
+void mf::WlrScreencopyManagerV1::capture_output(
+    struct wl_resource* frame,
+    int32_t overlay_cursor,
+    struct wl_resource* output)
+{
+    (void)overlay_cursor;
+    new WlrScreencopyFrameV1{frame, ctx, ctx->extents_of_output(output)};
+}
+
+void mf::WlrScreencopyManagerV1::capture_output_region(
+    struct wl_resource* frame,
+    int32_t overlay_cursor,
+    struct wl_resource* output,
+    int32_t x, int32_t y,
+    int32_t width, int32_t height)
+{
+    (void)overlay_cursor;
+    auto const intersection = geom::Rectangle{{x, y}, {width, height}}
+        .intersection_with(ctx->extents_of_output(output));
+    new WlrScreencopyFrameV1{frame, ctx, intersection};
+}
+
+mf::WlrScreencopyFrameV1::WlrScreencopyFrameV1(
+    wl_resource* resource,
+    std::shared_ptr<WlrScreencopyV1Ctx> const& ctx,
+    geometry::Rectangle const& area)
+    : wayland::WlrScreencopyFrameV1{resource, Version<3>()},
+      ctx{ctx},
+      area{area}
+{
+    send_buffer_event(
+        wayland::Shm::Format::argb8888,
+        area.size.width.as_uint32_t(),
+        area.size.height.as_uint32_t(),
+        area.size.width.as_uint32_t() * 4);
+    if (version_supports_buffer_done())
+    {
+        send_buffer_done_event();
+    }
+}
+
+void mf::WlrScreencopyFrameV1::copy(struct wl_resource* buffer)
+{
+    auto const graphics_buffer = ctx->allocator->buffer_from_shm(buffer, ctx->wayland_executor, [](){});
+    if (graphics_buffer->pixel_format() != mir_pixel_format_argb_8888)
+    {
+        BOOST_THROW_EXCEPTION(mw::ProtocolError(
+            resource,
+            Error::invalid_buffer,
+            "Invalid pixel format %d",
+            graphics_buffer->pixel_format()));
+    }
+    if (graphics_buffer->size() != area.size)
+    {
+        BOOST_THROW_EXCEPTION(mw::ProtocolError(
+            resource,
+            Error::invalid_buffer,
+            "Invalid buffer size %dx%d, should be %dx%d",
+            graphics_buffer->size().width.as_int(),
+            graphics_buffer->size().height.as_int(),
+            area.size.width.as_int(),
+            area.size.height.as_int()));
+    }
```

Comment:
Should also assert that `stride == width * 4`, like we've asked.

Reply:
Done